### PR TITLE
add support for tensorrt 8

### PR DESCRIPTION
**Pull request by hnyu**
Add support for TensorRT 8.6. Several important differences with TensorRT 10:
1. execute_async_v3 might not exist. It does for 8.6.2 (Jetson nano) but not for 8.6.1. Even it exists, the function might be buggy (all zero output). So for safety concern, we use v2 for TRT 8.
2. Some API differences when creating an engine.
3. A big trouble of TensorRT 8 is that its input/output args order might not respect that of the ONNX model! So we need to manually keep track of the correspondence when memcopying between host/device. For tensorrt 10, there is no such an issue as we can explicitly set the IO tensor mapping.

Since there is no backward compatibility, we need to implement two classes to handle them separately.

According to my test, TRT8 and TRT10 have the same performance. Also tested on Jetson nano and it passed the test case.

PS: onnxruntime can handle different TRT versions, so it always behaves normally.

### CODE REVIEW

File: alf/utils/tensorrt_utils.py [hunk 1]
```diff
@@ -450,6 +460,82 @@ def __call__(self, *args, **kwargs):
...
+
+    def __call__(self, *args, **kwargs):
+        flat_all_args = _dtype_conversions(alf.nest.flatten([args, kwargs]))
+
+        for i in range(len(flat_all_args)):
+            im = self._input_mem[i]
+            arg = flat_all_args[self._input_idx[i]]
+            cuda.memcpy_dtod_async(im,
+                                   arg.contiguous().data_ptr(),
+                                   self._get_bytes(arg), self._stream)
```
Comment (emailweixu) [inline alf/utils/tensorrt_utils.py:552]: synchronization is not needed here?

File: alf/utils/tensorrt_utils.py [hunk 2]
```diff
@@ -450,6 +460,82 @@ def __call__(self, *args, **kwargs):
...
+        self._output_mem = []
+        self._output_idx = []
+        self._bindings = []
+        # TRT8: This order might be different from the order of the onnx model!!
+        for i in range(engine.num_io_tensors):
+            name = engine.get_tensor_name(i)
+            shape = tuple(engine.get_tensor_shape(name))
+            idx = int(name.split('-')[1])
+            dtype = trt.nptype(engine.get_tensor_dtype(name))
+            host_mem = cuda.pagelocked_empty(shape, dtype)
```
Comment (emailweixu) [inline alf/utils/tensorrt_utils.py]: It seems that this memory allocation is just for calculating the size?
why not use `self._get_bytes()`

File: alf/utils/tensorrt_utils.py (hunk 1 above)
Comment (hnyu) [reply alf/utils/tensorrt_utils.py:552]: > synchronization is not needed here?

Actually this is still needed. I've added

File: alf/utils/tensorrt_utils.py (hunk 2 above)
Comment (hnyu) [reply alf/utils/tensorrt_utils.py]: > It seems that this memory allocation is just for calculating the size? why not use `self._get_bytes()`

Changed to use _get_bytes().

**Comment by hnyu** [comment]
I've changed the instruction to install TRT8 locally by default, to match the version on Jetson nano.

**Review by emailweixu — APPROVED** [review]

**Pull request merged**